lib/cnet/punt/punt_ether_kernel.c

```diff
@@ -0,0 +1,185 @@
+
+        int nb = pktdev_tx_burst(ctx->lport, mbufs, cnt);
+        if (nb == PKTDEV_ADMIN_STATE_DOWN)
+            CNE_WARN("Failed to send packets: %s\n", strerror(errno));
+    }
+}
+
+static uint16_t
+punt_ether_kernel_node_process(struct cne_graph *graph __cne_unused, struct cne_node *node,
+                               void **objs, uint16_t nb_objs)
+{
+    uint16_t n_left_from;
+    pktmbuf_t *mbufs[PREFETCH_CNT], **pkts;
+    int k;
+
+    pkts        = (pktmbuf_t **)objs;
+    n_left_from = nb_objs;
+
+    for (k = 0; k < PREFETCH_CNT && k < n_left_from; k++)
+        cne_prefetch0(pktmbuf_mtod_offset(pkts[k], void *, sizeof(struct cne_ether_hdr)));
+
+    while (n_left_from >= PREFETCH_CNT) {
+        /* Prefetch next-next mbufs */
+        if (likely(n_left_from > ((PREFETCH_CNT * 3) - 1))) {
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 0]);
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 1]);
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 2]);
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 3]);
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 4]);
+            cne_prefetch0(pkts[(PREFETCH_CNT * 2) + 5]);
+        }
+
+        /* Prefetch next mbuf data */
+        if (likely(n_left_from > ((PREFETCH_CNT * 2) - 1))) {
+            uint16_t pre = PREFETCH_CNT;
+
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 0], void *, pkts[pre + 0]->l2_len));
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 1], void *, pkts[pre + 1]->l2_len));
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 2], void *, pkts[pre + 2]->l2_len));
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 3], void *, pkts[pre + 3]->l2_len));
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 4], void *, pkts[pre + 4]->l2_len));
+            cne_prefetch0(pktmbuf_mtod_offset(pkts[pre + 5], void *, pkts[pre + 5]->l2_len));
+        }
+
+        memcpy(mbufs, pkts, (PREFETCH_CNT * sizeof(void *)));
+
+        pkts += PREFETCH_CNT;
+        n_left_from -= PREFETCH_CNT;
+
+        punt_ether_kernel_process_mbuf(node, mbufs, PREFETCH_CNT);
+    }
+
+    while (n_left_from > 0) {
+        mbufs[0] = pkts[0];
+
+        n_left_from--;
+        pkts++;
+
+        punt_ether_kernel_process_mbuf(node, mbufs, 1);
+    }
+
+    return nb_objs;
+}
+
+static int
+punt_ether_kernel_node_init(const struct cne_graph *graph __cne_unused, struct cne_node *node)
+{
+    punt_ether_kernel_node_ctx_t *ctx = (punt_ether_kernel_node_ctx_t *)node->ctx;
+
+    lport_cfg_t cfg = {0}; /**< CFG for tun/tap setup */
+    ctx->mmap       = mmap_alloc(DEFAULT_MBUF_COUNT, DEFAULT_MBUF_SIZE, MMAP_HUGEPAGE_4KB);
+    if (ctx->mmap == NULL)
+        cne_panic("Failed to mmap(%lu, %s) memory",
+                  (uint64_t)DEFAULT_MBUF_COUNT * (uint64_t)DEFAULT_MBUF_SIZE,
+                  mmap_name_by_type(MMAP_HUGEPAGE_4KB));
+
+    memset(&cfg, 0, sizeof(cfg));
+
+    strlcpy(cfg.name, TAP_NAME, sizeof(cfg.name));
```

Comment:
Could make the tap device name unique per cndp instance....

Reply:
I think it would be OK to have different names. I feel like I tried that, but not sure it was in this area.

I think with these changes we have two taps one tap at L2 and another after L4, is that correct?
It would be good to be able to see stats and multiple taps if possible.

Reply:
I think at L4 we weren't opening a tap? was it not just a socket that was being used to send the IPv4 packets down to the Kernel?

I guess we can try to add individual stats for the tap. I was just relying on gstats today to tell us how many packets went through the node and then by proxy how many went through the tap.

Reply:
We are using a socket to send packets to the kernel. As the code does not use tap here we can ignore the stats for now.

Are you ready for this PR to be reviewed and merged?

Reply:
> We are using a socket to send packets to the kernel. As the code does not use tap here we can ignore the stats for now.
> 
> Are you ready for this PR to be reviewed and merged?

So this "punt ether" node uses a raw tap. I tried the socket approach for L2 and had to create a tap device anyway because in order to send the L2 frame through the socket I needed an ifindex for an interface. Which is why here I opted to just use a raw tap device. We can certainly look at adding specific stats no probs. 

I will add a separate PR for the documentation around this node. 

The L3 punt node that already exists just uses a socket.